docs/proposals/koordlet/20230728-support-cold-memory-compute.md

```diff
@@ -0,0 +1,256 @@
+
+The part of memory that is allocated to application but probably unused for a while is called the "cold" memory. And the part of that in page cache is called cold page.
+
+**hot page**
+
+Relatively, the memory which may be reclaimable but would be reallocated shortly after is the "hot" memory. And the part of that in page cache is called hot page. 
+
+## Summary
+
+Implement the cold page info collecting and reporting. Based on that, we can incorporate cold memory compute into memory usage to support more fine-grained memory overcommitment in koordinator.
+
+## Motivation
+
+In general colocation scenarios, the resources that the high-priority (HP) applications have requested but not used are reclaimable to the low-priority (LP) applications to allocate. Resource reclaiming helps enhance the utilization of machines while increasing the risks of contentions. For memory resources, there are proportions of memory that are reusable on different levels. The free memory is allocatable to any process, while the page cache is potentially reusable after it is reclaimed by the system and no longer yields again. But  kernel does not proactively reclaim the page cache because for the better performance of the system, the kernel do not keep page cache idle and allocate the page cache to application as cache. To construct a more reliable and fine-grained memory overcommitment, the free and cold memory of the HP is preferred to reclaim to LP first. And if not required, the hot memory of HP should not be overcommitted, or the performance of HP can be affected. 
+
+In the Koordinator, we want to improve memory overcommitment by building the cold page reclaim ability, which can report the quantity of cold memory on the nodes and help assure that resource reclaiming like batch-memory are sound and efficient for the colocated applications.
+
+https://github.com/koordinator-sh/koordinator/issues/1178
+
+### Goals
+
+Define the API for the cold pages. Implement the cold page info collecting and reporting.
+
+### Non-Goals/Future Work
+
+Cold memory supports  scheduling optimization in koord-manager or koord-scheduler.
+
+## Proposal
+
+### Design
+
+#### Collect idle memory info and insert memory with hot page
+
+![image](../../images/support-cold-memory-1.svg)
+
+Add a coldPageCollector in collectorPlugins. coldPageCollector read cgroup file memory.idle_stat which is exported by kidled(Anolis kernel), kstaled(Google kernel)  or DAMON(amazon) that depends on the kernel. And memory.idle_stat includes cold page info in page cache and exists at each hierarchical of cgroup memory. After collecting cold page info, coldPageCollector will insert metric(such as NodeMemoryWithHotPageUsageMetric, PodMemoryWithHotPageUsageMetric, ContainerMemoryWithHotPageUsageMetric, NodeMemoryColdPageSizeMetric, PodMemoryColdPageSizeMetric, ContainerMemoryColdPageSizeMetric) into metriccache. 
+
+node memory usage: nodeMemWithHotPageUasge=MemTotal-MemFree+NodeMeMWithHotPageSize
```

Comment:
What is the definition of MemFree here, including the page cache part?

Reply:
That was a mistake.node memory usage: nodeMemWithHotPageUasge=MemTotal-MemFree-NodeMeMWithColdPage
MemTotal-MemFree includes all page cache and should minus cold page size. After that, mem usage includes hot page.

Reply:
Mem free not includes page cache.